### PR TITLE
Implement End-to-End Testing for Image-to-Form Processing Pipeline

### PR DESCRIPTION
**Pull request by Endlessflow**
#### Description:
This PR introduces a comprehensive end-to-end testing script for our image-to-form processing pipeline. The aim is to ensure the pipeline’s reliability and performance by facilitating seamless testing across various components, including OCR engines and LLM models.

#### Key Changes:
1. **End-to-End Testing Script (`e2e.py`)**:
   - **Pipeline Processing**: Developed a script to process input images through all combinations of OCR engines and LLM models.
   - **Directory Structure**: Established a structured directory setup under `e2e_tests` for input images, and outputs from OCR and LLM processes.
   - **Output Management**: Implemented checks to skip processing if output files already exist, optimizing the testing process and preventing redundant operations.
   - **Environment Management**: Improved handling of environment variables and enhanced logging for better configuration validation and error tracking.
   - **Error Handling**: Added detailed logging and error handling to provide insights during the processing flow.

2. **GPT Class Enhancements (`backend/gpt.py`)**:
   - **Debug Logging**: Added optional debug logging to track the timing and details of AzureOpenAI requests, aiding in performance monitoring.
   - **Request Parameter Handling**: Refactored the `GPT` class to improve parameter handling in an attempt to support models with varying feature availability, enhancing compatibility.
   
*Closes issue #45*

### CODE REVIEW

**Comment by Endlessflow** [comment]
@snakedye this is a draft mainly because I am not sure if `e2e.py` should stay in the root and if not where to put it. 

Otherwise, everything else is working and ready (except maybe Ollama support that needs to be discussed).

File: backend/gpt.py [hunk 1]
```diff
@@ -13,16 +15,32 @@ def __init__(self, api_endpoint, api_key, deployment="ailab-gpt-35-turbo-16k"):
             api_version = "2024-02-01",
         )
 
-    def generate_form(self, prompt):
+    def generate_form(self, prompt, DEBUG=False):
+        if DEBUG:
```
Comment (snakedye) [inline backend/gpt.py:19]: Everything under the `DEBUG` flag should go in the test file (./e2e.py)

File: backend/gpt.py [hunk 2]
```diff
@@ -13,16 +15,32 @@ def __init__(self, api_endpoint, api_key, deployment="ailab-gpt-35-turbo-16k"):
...
-        response = self.client.chat.completions.create(
-            model=self.model, # model = "deployment_name".
-            messages=[
-                {"role": "system", "content": system_prompt},
-                {"role": "user", "content": prompt}
-            ],
-            temperature=0,
-        )
+        
+        models_with_response_format = ["ailab-llm"] # List of models that support the response_format option
```
Comment (snakedye) [inline backend/gpt.py:27]: Set it as a constant at the beginning of the file.
```
MODELS_WITH_RESPONSE_FORMAT = ["ailab-llm"]
```

File: backend/gpt.py [hunk 3]
```diff
@@ -13,16 +15,32 @@ def __init__(self, api_endpoint, api_key, deployment="ailab-gpt-35-turbo-16k"):
...
+        request_params = {
+            "model": self.model, # model = "deployment_name".
+            "messages": messages,
+            "temperature": 0,
+        }
+        if self.model in models_with_response_format:
+            request_params["response_format"] = { "type": "json_object" }
+        response = self.client.chat.completions.create(**request_params)
+
+        if DEBUG:
```
Comment (snakedye) [inline backend/gpt.py:41]: Move to the test file.

File: e2e.py
Comment (snakedye) [inline e2e.py:1]: Find a more descriptive name for the file.

* test_e2e_pipeline.py
* generate_e2e_results.py

File: backend/gpt.py [hunk 4]
```diff
@@ -13,16 +15,32 @@ def __init__(self, api_endpoint, api_key, deployment="ailab-gpt-35-turbo-16k"):
             api_version = "2024-02-01",
         )
 
-    def generate_form(self, prompt):
+    def generate_form(self, prompt, DEBUG=False):
```
Comment (snakedye) [inline backend/gpt.py:18]: Remove the `DEBUG` flag.

**Review by snakedye — CHANGES_REQUESTED** [review]
GG, there's only this minor issue left.

**Comment by snakedye** [comment]
Don't forget to pass the checks 2 to 5. For the tests post a screenshot of the result of the test suite. Unfortunately the pipeline isn't fixed yet.

**Comment by Endlessflow** [comment]
Alright, it appears that everything is functioning as expected locally. Proceeding with the merge seems like a reasonable next step.

image

**Review by snakedye — APPROVED** [review]

**Pull request merged**